### PR TITLE
Allow switching tabs via context menu in the tabs area

### PR DESCRIPTION
**Pull request by Waqar144**
This change allows you to switch tabs via the context menu. The context
menu only shows up if you click on the empty area beside the tabs.

Signed-off-by: Waqar Ahmed <waqar.ahmed@kdab.com>

### CODE REVIEW

File: src/private/widgets/TabWidgetWidget.cpp [hunk 1]
```diff
@@ -201,3 +205,34 @@ void TabWidgetWidget::updateMargins()
...
+    auto tabAreaRect = tabBar->rect();
+    tabAreaRect.setWidth(this->width());
+    if (!tabAreaRect.contains(pos)) {
+        return;
+    }
+
+    QMenu menu(this);
+
+    for (int i = 0; i < count; ++i) {
+        menu.addAction(tabText(i), this, [this, i] {
```
Comment (iamsergio) [inline src/private/widgets/TabWidgetWidget.cpp:233]: disable the action for the current tab, as pressing it won't do anything

File: src/private/widgets/TabWidgetWidget_p.h [hunk 2]
```diff
@@ -68,6 +68,7 @@ class DOCKS_EXPORT TabWidgetWidget
 private:
     void updateMargins();
     void setupTabBarButtons();
+    void contextMenuRequested(QPoint pos);
```
Comment (iamsergio) [inline src/private/widgets/TabWidgetWidget_p.h:71]: make it protected and virtual, so users can create other context menus

File: src/private/widgets/TabWidgetWidget.cpp [hunk 3]
```diff
@@ -201,3 +205,34 @@ void TabWidgetWidget::updateMargins()
     m_cornerWidgetLayout->setContentsMargins(QMargins(0, 0, 2, 0) * factor);
     m_cornerWidgetLayout->setSpacing(int(2 * factor));
 }
+
+void TabWidgetWidget::contextMenuRequested(QPoint pos)
+{
+    auto *tabBar = QTabWidget::tabBar();
+    auto count = tabBar->count();
+    // We dont want context menu if there is only one tab
+    if (count <= 1) {
```
Comment (iamsergio) [inline src/private/widgets/TabWidgetWidget.cpp:214]: just count() instead ? no need for the temporary variables

**Review by iamsergio — CHANGES_REQUESTED** [review]

File: src/private/widgets/TabWidgetWidget.cpp (hunk 1 above)
Comment (Waqar144) [reply src/private/widgets/TabWidgetWidget.cpp]: Done

File: src/private/widgets/TabWidgetWidget.cpp (hunk 3 above)
Comment (Waqar144) [reply src/private/widgets/TabWidgetWidget.cpp]: Done

File: src/private/widgets/TabWidgetWidget_p.h (hunk 2 above)
Comment (Waqar144) [reply src/private/widgets/TabWidgetWidget_p.h]: Done

File: examples/dockwidgets/main.cpp [hunk 4]
```diff
@@ -135,15 +139,16 @@ int main(int argc, char **argv)
...
     QCommandLineOption noDropIndicators("no-drop-indicators", QCoreApplication::translate("main", "(internal) Don't use any drop indicators"));
 
+
     parser.addOption(noQtTool);
     parser.addOption(noParentForFloating);
     parser.addOption(nativeTitleBar);
     parser.addOption(noDropIndicators);
 
-# if defined(Q_OS_WIN)
+#if defined(Q_OS_WIN)
```
Comment (iamsergio) [inline examples/dockwidgets/main.cpp:148]: unrelated

**Review by iamsergio — CHANGES_REQUESTED** [review]

**Comment by iamsergio** [comment]
please run ctest , it should pass

I'm getting ASAN problems, try the dev preset

File: examples/dockwidgets/main.cpp (hunk 4 above)
Comment (Waqar144) [reply examples/dockwidgets/main.cpp]: reverted

**Comment by Waqar144** [comment]
> please run ctest , it should pass
> 
> I'm getting ASAN problems, try the dev preset

Were failing on master as well. However I included a fix

See `MainWindowBase.cpp` and `tst_docks` in the change list.

The height gets "0" as soon as the dockwidget is added to QTabWidget leading to a `divide by 0`, which seems weird. Not sure about the exact cause yet.

**Comment by iamsergio** [comment]
I've pushed the division by zero fix, since it's unrelated. Now checking why ASAN is failing

File: src/MainWindowBase.cpp [hunk 5]
```diff
@@ -788,4 +788,4 @@ QWidgetOrQuick *MainWindowBase::persistentCentralWidget() const
         return dw->widget();
 
     return nullptr;
-}
```
Comment (iamsergio) [inline src/MainWindowBase.cpp:791]: unrelated, I've removed it

**Comment by iamsergio** [comment]
this has been merged
